### PR TITLE
r/sagemaker_code_repository - new resource

### LINKED ISSUE #13817: Add support for AWS Sagemaker Code Repository resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Allow creation of a Git repository as a resource in Amazon SageMaker 

### New or Affected Resource(s)

* aws_sagemaker_code_repository
### References
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sagemaker-coderepository.html
https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateCodeRepository.html

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #13817

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_sagemaker_code_repository - new resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (26.45s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (35.76s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (36.04s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (84.66s)
```

`--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (1165.81s)`

### CODE REVIEW

File: aws/resource_aws_sagemaker_code_repository.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"code_repository_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 63),
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_code_repository.go:33]: Add additional validation:

> [Valid characters are a-z, A-Z, 0-9, and - (hyphen).](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateCodeRepository.html#API_CreateCodeRepository_RequestSyntax)

File: aws/resource_aws_sagemaker_code_repository.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	return resourceAwsSagemakerCodeRepositoryRead(d, meta)
+}
+
+func resourceAwsSagemakerCodeRepositoryRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	describeNotebookInput := &sagemaker.DescribeCodeRepositoryInput{
+		CodeRepositoryName: aws.String(d.Id()),
+	}
+	CodeRepository, err := conn.DescribeCodeRepository(describeNotebookInput)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_code_repository.go:91]:
```suggestion
	input := &sagemaker.DescribeCodeRepositoryInput{
		CodeRepositoryName: aws.String(d.Id()),
	}

	codeRepository, err := conn.DescribeCodeRepository(describeNotebookInput)
```

and change `CodeRepository` to `codeRepository` below.

File: aws/resource_aws_sagemaker_code_repository.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+	createOpts := &sagemaker.CreateCodeRepositoryInput{
+		CodeRepositoryName: aws.String(name),
+		GitConfig:          expandSagemakerCodeRepositoryGitConfig(d.Get("git_config").([]interface{})),
+	}
+
+	log.Printf("[DEBUG] sagemaker code repository create config: %#v", *createOpts)
+	_, err := conn.CreateCodeRepository(createOpts)
+	if err != nil {
+		return fmt.Errorf("Error creating Sagemaker code repository: %w", err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_code_repository.go:77]:
```suggestion
		return fmt.Errorf("Error creating SageMaker code repository: %w", err)
```

File: aws/resource_aws_sagemaker_code_repository.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+	}
+
+	if err := d.Set("code_repository_name", CodeRepository.CodeRepositoryName); err != nil {
+		return fmt.Errorf("error setting code_repository_name for sagemaker code repository (%s): %w", d.Id(), err)
+	}
+
+	if err := d.Set("arn", CodeRepository.CodeRepositoryArn); err != nil {
+		return fmt.Errorf("error setting arn for sagemaker code repository (%s): %w", d.Id(), err)
+	}
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_code_repository.go:108]: No need to check `err` for simple attribute values.

File: aws/resource_aws_sagemaker_code_repository.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+	input := &sagemaker.UpdateCodeRepositoryInput{
+		CodeRepositoryName: aws.String(d.Id()),
+		GitConfig:          expandSagemakerCodeRepositoryUpdateGitConfig(d.Get("git_config").([]interface{})),
+	}
+
+	log.Printf("[DEBUG] sagemaker code repository update config: %#v", *input)
+	_, err := conn.UpdateCodeRepository(input)
+	if err != nil {
+		return fmt.Errorf("Error updating Sagemaker code repository: %w", err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_code_repository.go:128]:
```suggestion
		return fmt.Errorf("error updating SageMaker code repository: %w", err)
```

File: aws/resource_aws_sagemaker_code_repository.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		CodeRepositoryName: aws.String(d.Id()),
+	}
+	CodeRepository, err := conn.DescribeCodeRepository(describeNotebookInput)
+	if err != nil {
+		if isAWSErr(err, "ValidationException", "Cannot find CodeRepository") {
+			d.SetId("")
+			log.Printf("[WARN] Unable to find SageMaker code repository (%s); removing from state", d.Id())
+			return nil
+		}
+		return fmt.Errorf("error finding sagemaker code repository (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_code_repository.go:98]:
```suggestion
		return fmt.Errorf("error reading SageMaker code repository (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_code_repository.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+	input := &sagemaker.DeleteCodeRepositoryInput{
+		CodeRepositoryName: aws.String(d.Id()),
+	}
+
+	if _, err := conn.DeleteCodeRepository(input); err != nil {
+		if isAWSErr(err, "ValidationException", "Cannot find CodeRepository") {
+			return nil
+		}
+		return fmt.Errorf("error trying to delete sagemaker code repository (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_code_repository.go:145]:
```suggestion
		return fmt.Errorf("error deleting SageMaker code repository (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_code_repository.go (hunk 2 above)
Comment (ewbankkit) [reply aws/resource_aws_sagemaker_code_repository.go:91]: Also, maybe move to an internal `finder` package (and reuse that function in `testAccCheckAWSSagemakerCodeRepositoryDestroy` and `testAccCheckAWSSagemakerCodeRepositoryExists`)?

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_ -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (17.27s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (23.32s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (29.28s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (46.38s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	46.431s
```

File: aws/resource_aws_sagemaker_code_repository.go (hunk 4 above)
Comment (DrFaust92) [reply aws/resource_aws_sagemaker_code_repository.go]: done

File: aws/resource_aws_sagemaker_code_repository.go (hunk 1 above)
Comment (DrFaust92) [reply aws/resource_aws_sagemaker_code_repository.go]: done

File: aws/resource_aws_sagemaker_code_repository.go (hunk 2 above)
Comment (DrFaust92) [reply aws/resource_aws_sagemaker_code_repository.go]: added

**Comment by DrFaust92** [comment]
addressed PR comments:

```
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (26.96s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (36.97s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (37.72s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (85.42s)
```

**Comment by DrFaust92** [comment]
Added test for code repo usage in sagemaker notebook instance:

```
--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (1165.81s)
```

File: aws/internal/service/sagemaker/finder/finder.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package finder
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sagemaker"
+)
+
+// FincCodeRepositoryByName returns the code repository corresponding to the specified name.
+// Returns nil if no code repository is found.
+func FincCodeRepositoryByName(conn *sagemaker.SageMaker, name string) (*sagemaker.DescribeCodeRepositoryOutput, error) {
```
Comment (ewbankkit) [inline aws/internal/service/sagemaker/finder/finder.go:10]:
```suggestion
// CodeRepositoryByName returns the code repository corresponding to the specified name.
// Returns nil if no code repository is found.
func CodeRepositoryByName(conn *sagemaker.SageMaker, name string) (*sagemaker.DescribeCodeRepositoryOutput, error) {
```

No need for the `Find` prefix as the package is named `finder`.

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_ -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (17.00s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (23.74s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (23.86s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (47.05s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	47.118s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic -timeout 120m
=== RUN   TestAccAWSSagemakerNotebookInstance_basic
=== PAUSE TestAccAWSSagemakerNotebookInstance_basic
=== RUN   TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== PAUSE TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== CONT  TestAccAWSSagemakerNotebookInstance_basic
=== CONT  TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
--- PASS: TestAccAWSSagemakerNotebookInstance_basic (336.17s)
--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (983.18s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	983.242s
```

**Comment by DrFaust92** [comment]
Rebased

**Comment by ewbankkit** [comment]
Verified acceptance tests after rebase:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerCodeRepository_\|TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_\|TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== RUN   TestAccAWSSagemakerNotebookInstance_basic
=== PAUSE TestAccAWSSagemakerNotebookInstance_basic
=== RUN   TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== PAUSE TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== CONT  TestAccAWSSagemakerNotebookInstance_basic
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
2020/10/29 14:34:13 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
2020/10/29 14:34:13 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
2020/10/29 14:34:13 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (19.23s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (24.93s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (25.02s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (50.26s)
--- PASS: TestAccAWSSagemakerNotebookInstance_basic (393.80s)
--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (1123.79s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1123.859s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @DrFaust92 

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_ -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (11.08s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (15.07s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (15.27s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (37.54s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	39.071s
```

Note that Code Repositories are not currently supported in the GovCloud region.

**Pull request merged**